lib/datadog/core/workers/interval_loop.rb

```diff
@@ -23,9 +23,11 @@ def perform(*args)
 
         def stop_loop
           mutex.synchronize do
-            return false unless run_loop?
+            run_loop = run_loop?
+            @run_loop = false # Always mark loop as stopped, in case worker thread hasn't started yet
+
+            return false unless run_loop
 
-            @run_loop = false
             shutdown.signal
           end
```

Comment:
I'll be honest -- after taking a second look at this modules, I don't think any of the changes are a good idea _at this time_.

That is because both `@mutex` and `@shutdown` are set in a racy way that is not guaranteed to actually be correct:

```ruby
        def mutex
          @mutex ||= Mutex.new
        end

        def shutdown
          @shutdown ||= ConditionVariable.new
        end
```

...so in its current state what we can say about this module is "in practice, it seems to work most of the time".

Being that we're operating on "works most of the time" mode, I very very much hesitate on touching anything at all regarding synchronization because that may end up triggering issues more often.

To be honest, I wonder if those are the source of the [Flaky CI issues](https://docs.google.com/spreadsheets/d/1kLX2bOxybFq4nUp8D9LEmRIs9khcG0Zo6HZahZwtGqA/edit#gid=1679005725) (Datadog-internal link, sorry to other community members!) we've seen for this module.